### PR TITLE
bugfix: just allow having use expr in `groupby` or `aggr`

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Part of #4556 .

having just use expr in groupby/aggr

# Rationale for this change

```sql
pg

-- create
CREATE TABLE EMPLOYEE (
  empId INTEGER PRIMARY KEY,
  name TEXT NOT NULL,
  dept TEXT NOT NULL
);

-- insert
INSERT INTO EMPLOYEE VALUES (0001, 'Clark', 'Sales');

-- fetch 
SELECT empid FROM EMPLOYEE having empid = 0;

psql:commands.sql:15: ERROR:  column "employee.empid" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT empid FROM EMPLOYEE having empid = 0;
               ^
```

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sql/group_by.rs [hunk 1]
```diff
@@ -194,7 +194,7 @@ async fn csv_query_group_by_and_having_and_where() -> Result<()> {
 async fn csv_query_having_without_group_by() -> Result<()> {
     let ctx = SessionContext::new();
     register_aggregate_csv(&ctx).await?;
-    let sql = "SELECT c1, c2, c3 FROM aggregate_test_100 HAVING c2 >= 4 AND c3 > 90";
+    let sql = "SELECT c1, c2, c3 FROM aggregate_test_100 where c2 >= 4 AND c3 > 90";
```
Comment (alamb) [inline datafusion/core/tests/sql/group_by.rs:197]: I think we can remove the whole test as we definitely have coverage for selecting with predicates in the where clause and the name of the test doesn't make sense `csv_query_having_without_group_by`

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -3641,10 +3632,11 @@ mod tests {
         let sql = "SELECT id, age
                    FROM person
                    HAVING age > 100 AND age < 200";
-        let expected = "Projection: person.id, person.age\
-                        \n  Filter: person.age > Int64(100) AND person.age < Int64(200)\
-                        \n    TableScan: person";
-        quick_test(sql, expected);
+        let err = logical_plan(sql).expect_err("query should have failed");
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:3635]: I recommend one test (that should also fail) where there is a GROUP clause but the having clause refers to an invalid column, such as

```sql
SELECT id, MAX(age)
FROM PERSON
GROUP BY id
-- first_name is a valid column but does not appear in the grouping output (group column nor aggregate)
HAVING first_name = 'M'
```

File: datafusion/sql/src/planner.rs [hunk 3]
```diff
@@ -1193,33 +1193,24 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                    check_columns_satisfy_exprs(
-                        &available_columns,
-                        &[having_expr.clone()],
-                        "HAVING clause references column(s) not provided by the select",
-                    )?;
+        let (plan, mut select_exprs_post_aggr, having_expr_post_aggr) = if !group_by_exprs
+            .is_empty()
+            || !aggr_exprs.is_empty()
+        {
+            self.aggregate(
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:1200]: I think we have to check here that the having expr only uses columns in the logical aggregation phase (group by columns or agg exprs)

**Review by alamb — COMMENTED** [review]
Thank you @jackwener 

I am not sure I would call this an "API break" -- more like a bugfix were DataFusion was previously allowing malformed SQL. 😆 

I had one negative case I think we may need to check but otherwise the code looks good to me 👍

File: datafusion/sql/src/planner.rs (hunk 3 above)
Comment (jackwener) [reply datafusion/sql/src/planner.rs:1200]: It will be check by `check_columns_satisfy_exprs`.

File: datafusion/sql/src/planner.rs (hunk 2 above)
Comment (jackwener) [reply datafusion/sql/src/planner.rs:3635]: Nice advice❤️
Added it.

**Comment by jackwener** [comment]
Thanks @alamb review, very helpful advice❤️.
has resolved them

File: datafusion/sql/src/planner.rs [hunk 4]
```diff
@@ -3646,7 +3646,20 @@ mod tests {
...
+
+    #[test]
+    fn select_with_having_refers_to_invalid_column() {
+        let sql = "SELECT id, MAX(age)
+                   FROM person
+                   GROUP BY id
+                   HAVING first_name = 'M'";
+        let err = logical_plan(sql).expect_err("query should have failed");
+        assert_eq!(
+            "Plan(\"HAVING clause references non-aggregate values: Expression person.first_name could not be resolved from available columns: person.id, MAX(person.age)\")",
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:3662]: Very nice!

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @jackwener

**Comment by mingmwang** [comment]
@jackwener Could you please add a UT to cover this case ?

`SELECT id  FROM person GROUP BY id  ORDER BY MAX(age);`

Looks like this is allowed in PG and SparkSQL.

**Pull request merged**